### PR TITLE
Increase visibility for retries

### PR DESCRIPTION
**Pull request by ThomasObenaus**

### CODE REVIEW

File: cosmos.go [hunk 1]
```diff
@@ -356,7 +373,7 @@ func (c *cosmosImpl) executeAsync(query string, asyncResponses *[]interfaces.Asy
 	if err := c.pool.ExecuteAsync(query, intermediateChannel); err != nil {
 		return nil, err
 	}
-	errorCallback(err)
+	defer errorCallback(err)
```
Comment (daniel-meyer-dme) [inline cosmos.go:376]: Should not be deferred, as this should return the error of the first call to executeAsync otherwise it is not async anymore.
```suggestion
	errorCallback(err)
```

File: cosmos.go [hunk 2]
```diff
@@ -266,23 +266,31 @@ func (c *cosmosImpl) ExecuteWithBindings(query string, bindings, rebindings map[
...
 	for tryCount = 0; tryCount < maxTries; tryCount++ {
-
 		responses, err = executeRequest()
+		if err != nil {
+			metrics.requestErrorsTotal.Inc()
+			return nil, errors.Wrap(err, "executing request in retry loop")
+		}
 
-		updateRequestMetrics(responses, c.metrics)
+		isARetry := tryCount > 0
```
Comment (daniel-meyer-dme) [inline cosmos.go:290]:
```suggestion
		isARetry := tryCount > 1
```
TryCount is 0-based in the for-loop

File: cosmos_test.go [hunk 3]
```diff
@@ -1238,11 +1239,9 @@ func TestWaitForRetry(t *testing.T) {
 }
 
 func TestWaitForRetry_Abort(t *testing.T) {
-
 	// GIVEN
-	waitTime := time.Millisecond * 20
-	stop := make(chan bool)
-	defer close(stop)
+	waitTime := time.Millisecond * 100
```
Comment (daniel-meyer-dme) [inline cosmos_test.go:1243]: what is the reason for the changed durations?

File: cosmos_test.go [hunk 4]
```diff
@@ -1254,7 +1253,8 @@ func TestWaitForRetry_Abort(t *testing.T) {
 		waitDone = waitForRetry(waitTime, stop)
 		called = true
 	}()
-	stop <- true
+	time.Sleep(time.Millisecond * 20)
```
Comment (daniel-meyer-dme) [inline cosmos_test.go:1256]: what is the reason for the changed durations?

File: cosmos.go [hunk 5]
```diff
@@ -295,40 +303,49 @@ func (c *cosmosImpl) retryLoop(executeRequest retryFun) (responses []interfaces.
...
 	}
 
 	return responses, err
 }
 
-func (c *cosmosImpl) handleTimeout(done <-chan bool) (timedOutChan <-chan bool) {
-
-	timeoutReachedChan := make(chan bool)
+func handleTimeout(done <-chan struct{}, retryTimeout time.Duration, logger zerolog.Logger) (timedOutChan <-chan struct{}) {
+	timeoutReachedChan := make(chan struct{})
```
Comment (daniel-meyer-dme) [inline cosmos.go:331]: This changes the results of this producer, as the channel closed in both cases (timedOut, done) and thus the caller can't determine which situation occurred.

File: cosmos.go [hunk 6]
```diff
@@ -295,40 +303,49 @@ func (c *cosmosImpl) retryLoop(executeRequest retryFun) (responses []interfaces.
...
 			}
 		}
 
 		// Timeout check in case no waiting is required
 		select {
 		case <-timeoutReachedChan:
 			// we stop here and return what we got so far
-			return responses, err
+			metrics.requestRetryTimeoutsTotal.Inc()
+			return nil, fmt.Errorf("timed out while doing a retry (timeout=%s)", retryTimeout)
```
Comment (daniel-meyer-dme) [inline cosmos.go:320]: not sure if this is the better solution, the first answer received would still be more valuable, even if it failed due to throttling or whatever

File: cosmos_test.go [hunk 7]
```diff
@@ -1268,23 +1268,20 @@ func TestWaitForRetry_Abort(t *testing.T) {
...
 	// WHEN
 	go func() {
-		timedOutChan := cosmos.handleTimeout(done)
+		timedOutChan := handleTimeout(done, retryTimeout, zerolog.Nop())
 
-		timeOut := <-timedOutChan
+		<-timedOutChan
 		mu.Lock()
-		timedOut = timeOut
+		timedOut = true
```
Comment (daniel-meyer-dme) [inline cosmos_test.go:1284]: now timedOut is always true whether something was sent on the timedOutChan or not

File: cosmos_test.go [hunk 8]
```diff
@@ -1302,23 +1299,20 @@ func TestHandleTimeout(t *testing.T) {
...
 	// WHEN
 	go func() {
-		timedOutChan := cosmos.handleTimeout(done)
+		timedOutChan := handleTimeout(done, retryTimeout, zerolog.Nop())
 
-		for isTimedOut := range timedOutChan {
+		for range timedOutChan {
 			mu.Lock()
-			timedOut = isTimedOut
+			timedOut = true
```
Comment (daniel-meyer-dme) [inline cosmos_test.go:1315]: now timedOut is always true whether something was sent on the timedOutChan or not

File: cosmos_test.go [hunk 9]
```diff
@@ -1337,3 +1331,163 @@ func TestHandleTimeout_Abort(t *testing.T) {
...
+	responses, err := retryLoop(retryFn, 1, time.Second, metrics, zerolog.Nop())
+
+	// THEN
+	assert.Error(t, err)
+	assert.Empty(t, responses)
+}
+
+func TestHandleRetryLoop_FailWhileCallWasTooLong(t *testing.T) {
+	// In this test the call to cosmos takes 1s
+	// At the same time we defined to wait for 200ms at max.
```
Comment (daniel-meyer-dme) [inline cosmos_test.go:1437]: Where are those numbers defined?

File: cosmos.go [hunk 10]
```diff
@@ -266,23 +266,31 @@ func (c *cosmosImpl) ExecuteWithBindings(query string, bindings, rebindings map[
...
+			metrics.requestErrorsTotal.Inc()
+			return nil, errors.Wrap(err, "executing request in retry loop")
+		}
 
-		updateRequestMetrics(responses, c.metrics)
+		isARetry := tryCount > 0
+		updateRequestMetrics(responses, metrics, isARetry)
 
-		if !shouldRetry || err != nil {
+		if !shouldRetry {
```
Comment (daniel-meyer-dme) [inline cosmos.go:293]: Why was this removed? The retry feature should not handle errors, but only azure defined retries

**Review by daniel-meyer-dme — CHANGES_REQUESTED** [review]

File: cosmos.go (hunk 6 above)
Comment (ThomasObenaus) [reply cosmos.go:320]: - will move the logmessage 
- won't return an error 
- will return the responses

File: cosmos.go (hunk 1 above)
Comment (ThomasObenaus) [reply cosmos.go:376]: done

File: cosmos.go (hunk 5 above)
Comment (ThomasObenaus) [reply cosmos.go:331]: will check to switch back to bool chan

File: cosmos.go (hunk 10 above)
Comment (ThomasObenaus) [reply cosmos.go:293]: clarified

**Review by daniel-meyer-dme — APPROVED** [review]
